### DISCUSSION TITLE
CustomDocumentLoader, how to return an empty document(only add objects to module's context, no code needed)

### DISCUSSION THREAD

**Discussion by nextfool** [ClearFoundry/ClearScript#322] [General]
Hi @ClearScriptLib ,

I don't know if I am on the right path, but here is what I intend to do:
I want to expose some api/objects(implemented with c# of course) with the "import syntax"
import {SomeApi} form "@systemApi/SomeApi",
so I created a custom document loader, whenever the "specifier" matches my system api uri format, I pass a "DocumentContextCallback" which create a "context" (with corresponding javascript api exposed) to the default loader's "LoadDocumentAsync".
The problem is that, I don't have any "runnable"  javascript code for the "module", so I am thinking of returning an empty document.

Please see the screenshot bellow.

![image](https://user-images.githubusercontent.com/2167369/141791243-e884aced-9d96-48fa-828f-e49460d290b3.png)

Can you please tell me whether it is the correct way or you have any other advice?

Thank you.

**Comment by ClearScriptLib** [comment]
Hi @nextfool,

The problem with your approach is that a module's context is only accessible within the module, so an empty module can serve no purpose. It needs to export something in order to be useful. The context, accessible in ES6 modules as [`import.meta`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import.meta), allows the host to give the module exclusive access to a protected resource such as a host object or type.

Here's an example that sets up a minimal "@system/console" module. Note that you can use the [system document](https://clearscript.clearfoundry.net/Reference/html/Overload_Microsoft_ClearScript_DocumentSettings_AddSystemDocument.htm) feature instead of a custom document loader for this purpose:

```C#
const string consoleCode = @"
    export function log() {
        import.meta.Console.WriteLine(...arguments);
    }
";
engine.DocumentSettings.AddSystemDocument(
    "@system/console",
    ModuleCategory.Standard,
    consoleCode,
    _ => new Dictionary<string, object> {
        { "Console", typeof(Console).ToHostType() }
    }
);
```

With this setup, you can do this:

```C#
engine.Execute(new DocumentInfo { Category = ModuleCategory.Standard }, @"
    import * as console from '@system/console';
    console.log('Hello, world! {0} is the current time.', new Date().toLocaleTimeString());
");
```

Please let us know if you have additional questions.

Thanks!
